SCClassLibrary/Common/Audio/BasicOpsUGen.sc

```diff
@@ -202,13 +203,15 @@ BinaryOpUGen : BasicOpUGen {
 			and: { a.descendants.size == 1 }})
 		{
 			if (MulAdd.canBeMulAdd(a.inputs[0], a.inputs[1], b)) {
+				if(a === b) { ^nil };
```

Comment:
Come to think of it, all of these tests can be the first line in their respective methods. If a === b, there is no need to do any more complex checks -- check the simple thing first and bail out if possible.

```
	optimizeToMulAdd {
		var a, b, replacement;
		#a, b = inputs;

		if(a === b) { ^nil };
		if (a.isKindOf(BinaryOpUGen) and: { a.operator == '*'
... snip
```

Reply:
Yes, now that I see it again I think is logically the same for all cases to put it before the other checks, I was following optimizeToSum3 but it makes not much sense for these other methods.